src/connectors/tests/clickhouse.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2021, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// As we only have a sink implementation, we'll put everything here. Everything
+// will eventually follow a structure similar to the s3 connector.
+
+use crate::errors::{Error, Result};
+
+use std::time::Duration;
+
+use async_std::stream::StreamExt;
+use clickhouse_rs::Pool;
+use signal_hook::consts::{SIGINT, SIGQUIT, SIGTERM};
+use signal_hook_async_std::Signals;
+use testcontainers::{clients, core::Port, images::generic::GenericImage, Docker, RunArgs};
+use tremor_common::ports::IN;
+use tremor_pipeline::{Event, EventId};
+use tremor_script::literal;
+
+use super::ConnectorHarness;
+
+const CONTAINER_NAME: &str = "clickhouse/clickhouse-server";
+const CONTAINER_VERSION: &str = "22.3.3.44";
+const SERVER_PORT: u16 = 9000;
+
+const DB_HOST: &str = "127.0.0.1";
+
+#[async_std::test]
+async fn simple_insertion() -> Result<()> {
+    let _ = env_logger::try_init();
+
+    let docker = clients::Cli::default();
+    let image = GenericImage::new(format!("{CONTAINER_NAME}:{CONTAINER_VERSION}"));
+
+    // We want to access the container from the host, so we need to make the
+    // corresponding port available.
+    let port_to_expose = Port {
+        local: SERVER_PORT,
```

Comment:
just for the sake of your sanity, use our tooling in `free_port::find_free_tcp_port()` or use `#[serial_test]`, otherwise this will haunt you (or maybe us later on) forever. ;-)

Reply:
You might even leave this out completely and the OS will pick a free port for you :)

Reply:
Thanks for the tip. That explains why I had to stop my local, non-docker clickhouse server :sob:

Reply:
This should be fixed by commit [2964597](https://github.com/tremor-rs/tremor-runtime/pull/1538/commits/296459726f49053020aaf3ba33178634be8a2a53). Feel free to reopen if some changes are still needed :)